### ISSUE TITLE
What does PPS score?

### ISSUE THREAD

**Issue by deeepwin** [8080labs/ppscore#62]
Hi Florian,

first - I think the predictive power score is a great idea. Thank you very much for that. However, I am struggling to understand what PPS actually scores.

In one of the comments you have written, that the technical interpretation of the PPS is: the percentage of model improvement potential that the feature adds when the current naive base model is compared to a perfect deterministic model.

This I understand. But what means predictive power? My understanding is, in the context of this score, the predictive power of the model (in this case a decision tree) predicting y given X. Particularly, how much better the decision tree is compared to the the naive base model.

However, I am looking for a score, that tells me how good X is to predict y. As more "patterns" or generally "information" X has, as better X correlates in some form to y. The problem is: How well X can predict y. I am aware that there are entropy based scores such as MIC or MI or regular correlation such as Pearson. All are not model based.

I am just not sure, If I can really use PPS and compare with the other methods? 

My line of thoughts are, that PPS depends on how well the model fits the data (overfits, underfits or perfect fit).  The error/loss from the model will be different. PPS will be different, but X has not changed at all. In most cases, if model fits well, there is some information in X. If there is nothing in X, model won't be able to fit anything. Such as, if X is completely uncorrelated to y. It means PPS is a total , summing up predictive power of X and the predictive power of the model itself. But, these two powers could possibly cancel each out. In such case, X would have large predictive power, but model not due to bad fitting, then PPS would be 0, defying its purpose.

I would very much appreciate your feedback on that conclusion. Thank you.

**Comment by FlorianWetschoreck** [comment]
Thank you for starting the discussion!

> But what means predictive power?
As you mentioned above: In the context of PPS, predictive power is defined as the percentage of model improvement potential that the feature adds when the current naive base model is compared to a perfect deterministic model.

And you are right that this depends on the performance of the model. In case the model overfits or underfits, this impacts PPS. Thus, decision trees were chosen because they usually have a good, balanced performance. If you find a better default model, then please let us know.

> I am looking for a score, that tells me how good X is to predict y

Well, I think data itself cannot predict other data. You always need a model - an interpretation of how you will map from x to y. Methods such as MIC, MI, correlation etc also use an implicit model or they measure something different but not predictability.

> The problem is: How well X can predict y

Indeed. I think that you can only answer this after the fact e.g. I found 3 models and the best of them has a certain score. I don't know if I might find a better model in the future. How long will you try to find something better? I see this as the "halting problem".

In the end, the question is:
Which task are you trying to accomplish? How do you hope the PPS will help you doing this? Which of your expectations are not satisfied?

What do you think?

**Comment by deeepwin** [comment]
Thank you for picking up the discussion. I very much appreciate it!

I completely agree, a decision tree is a good choice. It is a structure with manageable complexity and covers non-linear mapping. I also fully agree on your statement about explicit and implicit model. This is a good point. A correlation, like a Pearson correlation is a mapping respectively implicit model as well.

However, one aspect I would like to throw in. Mutual information MI is based on information theory. MI like decision tree uses entropy. To decide on the splitting value in a decision tree information gain is used, which is entropy between parent and child nodes. 

Information theory origin is about communication channel and understanding, if information has been lost when transmitting data (Shannon). One could think about the same way for x and y data. If a pattern exists in x and if that pattern would reoccur in y, possibly slightly modified and with more noise, an information theory metric should be able to detect it.

My original idea before discovering PPS, was to use the information gain IG the decision tree has a metric of predictive power, instead of the loss. Randomly related x any would have almost zero IG. x and y with some dependency would have more than zero IG. Unfortunately, in a decision tree, each level calculates its own IG. It cannot be added up, at least not to my understanding.

What do you think about the information theory approach?

~~~~
Indeed. I think that you can only answer this after the fact e.g. I found 3 models and the best of them has a certain score. I don't know if I might find a better model in the future. How long will you try to find something better? I see this as the "halting problem".

In the end, the question is:
 Which task are you trying to accomplish? How do you hope the PPS will help you doing this? Which of your expectations are not satisfied?
~~~~

This is the price question. I see it this way, PPS uses decision trees, you can use another model. But, the score will always be related to that model.  If you use a similar model in your machine learning production model, you have a easy metric with PPS to approximate and possible pre-select features.  In that case, the question is more, if pre-selection makes sense at all, since the production model might be already inexpensive in regard to computing power.

However, if you are using a more complex production model such as a deep learning neural network, PPS would be very nice to pre-select feature (x) with high predictability. But, here we have the problem, that decision tree and production model might differ that much, that the PPS score mismatches with what the production model detects. On the other hand finding features with high predictive power directly with production model is very expensive. 

Both approaches are sub optimal. Then it goes back the the question you stated above.

**Comment by FlorianWetschoreck** [comment]
Thank you for your reply. I will give this some more thought. Also, I want to research some questions about information theory etc before I answer.

**Comment by deeepwin** [comment]
Think that can be closed. Thank you.
